implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java

```diff
@@ -0,0 +1,247 @@
+        }
+    }
+
+    private void tryProfiles(final URI uri, final ConfigSourceContext context, final List<ConfigSource> configSources) {
+        final List<String> profiles = context.getProfiles();
+        for (int i = profiles.size() - 1; i >= 0; i--) {
+            final int lastOrdinal = configSources.get(configSources.size() - 1).getOrdinal();
+            final URL profileToFileName = toURL(addProfileToFileName(uri, profiles.get(i)));
+            consumeAsPath(profileToFileName, path -> addProfileConfigSource(profileToFileName, lastOrdinal + 1, configSources));
+        }
+    }
+
+    private static URL toURL(final URI uri) {
+        try {
+            return uri.toURL();
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    private void addConfigSource(final URI uri, final List<ConfigSource> configSources) {
+        addConfigSource(toURL(uri), configSources);
+    }
+
+    private void addConfigSource(final URL url, final List<ConfigSource> configSources) {
+        try {
+            final ConfigSource configSource = loadConfigSource(url);
+            if (!configSource.getProperties().isEmpty()) {
+                configSources.add(configSource);
+            }
+        } catch (IOException e) {
+            throw ConfigMessages.msg.failedToLoadResource(e);
+        }
+    }
+
+    private void addProfileConfigSource(final URL url, final int ordinal, final List<ConfigSource> configSources) {
+        try {
+            final ConfigSource configSource = loadConfigSource(url, ordinal);
+            if (!configSource.getProperties().isEmpty()) {
+                configSources.add(configSource);
+            }
+        } catch (FileNotFoundException e) {
+            // It is ok to not find the resource here, because it is an optional profile resource.
+        } catch (IOException e) {
+            throw ConfigMessages.msg.failedToLoadResource(e);
+        }
+    }
+
+    private boolean validExtension(final Path fileName) {
+        return validExtension(fileName.getFileName().toString());
+    }
+
+    private boolean validExtension(final String resourceName) {
+        for (String s : getFileExtensions()) {
+            if (resourceName.endsWith(s)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static URI addProfileToFileName(final URI uri, final String profile) {
+        final String fileName = uri.getPath() != null ? uri.getPath() : uri.getRawSchemeSpecificPart();
+        assert fileName != null;
+
+        final int dot = fileName.lastIndexOf(".");
+        final String fileNameProfile;
+        if (dot != -1) {
+            fileNameProfile = fileName.substring(0, dot) + "-" + profile + fileName.substring(dot);
+        } else {
+            fileNameProfile = fileName + "-" + profile;
+        }
+
+        return URI.create(uri.toString().replace(fileName, fileNameProfile));
```

Comment:
This doesn't really solve any of the quoting issues and in fact can blow up in various humorous ways.  You probably want something safer like this:
```suggestion
        return new URI(uri.getScheme(), uri.getAuthority(), uri.getHost(), uri.getPort(), fileNameProfile, uri.getQuery(), uri.getFragment());
```

Reply:
Forgot to add a comment here or a TODO for us to discuss. I've just added this to move forward.

Your suggestion was my first implementation, but it doesn't work for the `jar` scheme, because the path is stored in the  `schemeSpecificPart`, so we need some additional logic to cover that case.